Parallel codegen regression

```
Global is external, but doesn't have external or weak linkage!
void (%"1.core::option::Option<(&'static mut u32, &'static u32)>"*, i64, i32*)* @"_ZN4iter30Zip$LT$A$C$$u20$B$GT$.Iterator4next12closure.5367E"
invalid linkage type for function declaration
void (%"1.core::option::Option<(&'static mut u32, &'static u32)>"*, i64, i32*)* @"_ZN4iter30Zip$LT$A$C$$u20$B$GT$.Iterator4next12closure.5367E"
Global is external, but doesn't have external or weak linkage!
void (%"1.core::option::Option<(&'static mut u32, &'static u32)>"*, i64, i32*)* @"_ZN4iter30Zip$LT$A$C$$u20$B$GT$.Iterator4next12closure.5372E"
invalid linkage type for function declaration
void (%"1.core::option::Option<(&'static mut u32, &'static u32)>"*, i64, i32*)* @"_ZN4iter30Zip$LT$A$C$$u20$B$GT$.Iterator4next12closure.5372E"
LLVM ERROR: Broken module found, compilation aborted!
/home/ncameron/rust2/mk/target.mk:171: recipe for target 'x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/stamp.rand' failed
make: *** [x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/stamp.rand] Error 1
make: *** Waiting for unfinished jobs....
Global is external, but doesn't have external or weak linkage!
i32 (%closure.2, i32)* @"_ZN3str28Chars$LT$$u27$a$GT$.Iterator4next13closure.16349E"
invalid linkage type for function declaration
i32 (%closure.2, i32)* @"_ZN3str28Chars$LT$$u27$a$GT$.Iterator4next13closure.16349E"
Global is external, but doesn't have external or weak linkage!
Global is external, but doesn't have external or weak linkage!
i64 (i64, i32)* @"_ZN4iter30Map$LT$I$C$$u20$F$GT$.Iterator4next13closure.16444E"
invalid linkage type for function declaration
i64 (i64, i32)* @"_ZN4iter30Map$LT$I$C$$u20$F$GT$.Iterator4next13closure.16444E"
Global is external, but doesn't have external or weak linkage!
i32 (%closure.2, i32)* @"_ZN3str28Chars$LT$$u27$a$GT$.Iterator4next13closure.16367E"
invalid linkage type for function declaration
i64 (%closure.6*, i64, i64)* @_ZN4iter8Iterator3sum13closure.16450E
invalid linkage type for function declaration
i64 (%closure.6*, i64, i64)* @_ZN4iter8Iterator3sum13closure.16450E
Global is external, but doesn't have external or weak linkage!
Global is external, but doesn't have external or weak linkage!i32 (%closure.2, i32)* @"_ZN3st
r28Chars$LT$$u27$a$GT$.Iterator4next13closure.16367E"
Global is external, but doesn't have external or weak linkage!
i32 (%closure.11, i32)* @"_ZN3str39Chars$LT$$u27$a$GT$.DoubleEndedIterator9next_back13closure.16531E"
invalid linkage type for function declaration
i64 (i64, i32)* @"_ZN4iter30Map$LT$I$C$$u20$F$GT$.Iterator4nexti13closure32 (.164%55E"closure.11
, invalid linkage type for function declarationi32
)* @"_ZN3str39Chars$LT$$u27$a$GT$.DoubleEndedIterator9next_back13closure.16531E"
LLVM ERROR: Broken module found, compilation aborted!
pure virtual method called
terminate called without an active exception
```

--- a/src/librustc/middle/ty.rs
+++ b/src/librustc/middle/ty.rs
@@ -3283,6 +3283,7 @@ impl FlagComputation {
             }
 
             &TyClosure(_, substs) => {
+                self.add_flags(TypeFlags::HAS_TY_CLOSURE);
                 self.add_flags(TypeFlags::HAS_LOCAL_NAMES);
                 self.add_substs(substs);
             }

--- a/src/test/run-pass/parallel-codegen-closures.rs
+++ b/src/test/run-pass/parallel-codegen-closures.rs
@@ -1,0 +1,34 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Tests parallel codegen - this can fail if the symbol for the anonymous
+// closure in `sum` pollutes the second codegen unit from the first.
+
+// compile-flags: -C codegen_units=2
+
+#![feature(core)]
+#![feature(iter_arith)]
+
+mod a {
+    fn foo() {
+        let x = ["a", "bob", "c"];
+        let len: usize = x.iter().map(|s| s.len()).sum();
+    }
+}
+
+mod b {
+    fn bar() {
+        let x = ["a", "bob", "c"];
+        let len: usize = x.iter().map(|s| s.len()).sum();
+    }
+}
+
+fn main() {
+}

